server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,668 @@
+        starNode.isStarNode = true;
+        starNode.startDocId = numStarTreeDocs;
+        Iterator<StarTreeDocument> starTreeDocumentIterator = generateStarTreeDocumentsForStarNode(startDocId, endDocId, dimensionId);
+        while (starTreeDocumentIterator.hasNext()) {
+            appendToStarTree(starTreeDocumentIterator.next());
+        }
+        starNode.endDocId = numStarTreeDocs;
+        return starNode;
+    }
+
+    /**
+     * Returns aggregated star-tree document
+     *
+     * @param node star-tree node
+     * @return aggregated star-tree documents
+     * @throws IOException throws an exception upon failing to create new aggregated docs based on star tree
+     */
+    private StarTreeDocument createAggregatedDocs(TreeNode node) throws IOException {
+        StarTreeDocument aggregatedStarTreeDocument = null;
+        if (node.children == null) {
+
+            // For leaf node
+            if (node.startDocId == node.endDocId - 1) {
+                // If it has only one document, use it as the aggregated document
+                aggregatedStarTreeDocument = getStarTreeDocument(node.startDocId);
+                node.aggregatedDocId = node.startDocId;
+            } else {
+                // If it has multiple documents, aggregate all of them
+                for (int i = node.startDocId; i < node.endDocId; i++) {
+                    aggregatedStarTreeDocument = reduceStarTreeDocuments(aggregatedStarTreeDocument, getStarTreeDocument(i));
+                }
+                if (null == aggregatedStarTreeDocument) {
+                    throw new IllegalStateException("aggregated star-tree document is null after reducing the documents");
+                }
+                for (int i = node.dimensionId + 1; i < numDimensions; i++) {
+                    aggregatedStarTreeDocument.dimensions[i] = Long.valueOf(STAR_IN_DOC_VALUES_INDEX);
+                }
+                node.aggregatedDocId = numStarTreeDocs;
+                appendToStarTree(aggregatedStarTreeDocument);
+            }
+        } else {
+            // For non-leaf node
+            if (node.children.containsKey((long) ALL)) {
+                // If it has star child, use the star child aggregated document directly
+                for (TreeNode child : node.children.values()) {
+                    if (child.isStarNode) {
+                        aggregatedStarTreeDocument = createAggregatedDocs(child);
+                        node.aggregatedDocId = child.aggregatedDocId;
+                    } else {
+                        createAggregatedDocs(child);
+                    }
+                }
+            } else {
+                // If no star child exists, aggregate all aggregated documents from non-star children
+                if (node.children.values().size() == 1) {
+                    for (TreeNode child : node.children.values()) {
+                        aggregatedStarTreeDocument = createAggregatedDocs(child);
```

Comment:
```
aggregatedStarTreeDocument = reduceStarTreeDocuments(aggregatedStarTreeDocument, createAggregatedDocs(child));
```
Lets do this instead , just to be safe. We can take this up in next PR

Reply:
actually we can skip this comment since aggregatedStarTreeDocument is not used in this condition. This looks good. Reduce is of no use.